src/Middleware/HttpLogging/src/FileLoggerProcessor.cs

```diff
@@ -0,0 +1,163 @@
+                    {
+                        await WriteMessagesAsync(_currentBatch);
+                    }
+                    catch
+                    {
+                        // ignored
+                    }
+
+                    _currentBatch.Clear();
+                }
+            }
+        }
+
+        internal async Task WriteMessagesAsync(IEnumerable<LogMessage> messages)
+        {
+            Directory.CreateDirectory(_path);
+
+            // Files are grouped by day, and written up to _maxFileSize before rolling to a new file
+            foreach (var group in messages.GroupBy(GetGrouping))
+            {
+                var fullName = GetFullName(group.Key);
+                var fileInfo = new FileInfo(fullName);
+                var streamWriter = File.AppendText(fullName);
+
+                foreach (var item in group)
+                {
+                    // Roll to new file if _maxFileSize is reached
+                    // _maxFileSize could be less than the length of the file header - in that case we still write the first log message before rolling
+                    if (fileInfo.Length > _maxFileSize)
+                    {
+                        _fileNumber++;
+                        fullName = GetFullName(group.Key);
+                        fileInfo = new FileInfo(fullName);
+                        streamWriter.Dispose();
+                        streamWriter = File.AppendText(fullName);
+                    }
+                    if (fileInfo.Length == 0)
+                    {
+                        await OnFirstWrite(streamWriter);
+                    }
+
+                    await WriteMessageAsync(item.Message, streamWriter);
+                }
+                streamWriter.Dispose();
+            }
+
+            RollFiles();
+        }
+
+        internal async Task WriteMessageAsync(string message, StreamWriter streamWriter)
+        {
+            await streamWriter.WriteLineAsync(message);
+            await streamWriter.FlushAsync();
+        }
+
+        private void RollFiles()
+        {
+            if (_maxRetainedFiles > 0)
+            {
+                var files = new DirectoryInfo(_path)
+                    .GetFiles(_fileName + "*")
+                    .OrderByDescending(f => f.Name)
+                    .Skip(_maxRetainedFiles.Value);
+
+                foreach (var item in files)
+                {
+                    item.Delete();
+                }
+            }
+        }
+
+        public void Dispose()
+        {
+            _cancellationTokenSource.Cancel();
+            _messageQueue.CompleteAdding();
+            _outputTask.Wait();
+        }
+
+        private string GetFullName((int Year, int Month, int Day) group)
+        {
+            return Path.Combine(_path, $"{_fileName}{group.Year:0000}{group.Month:00}{group.Day:00}{_fileNumber:00}.txt");
+        }
+
+        private (int Year, int Month, int Day) GetGrouping(LogMessage message)
```

Comment:
Is it possible to use the new `DateOnly` type?
```suggestion
        private DateOnly GetGrouping(LogMessage message)
```